src/test/scala/org/pac4j/http4s/SessionSpec.scala

```diff
@@ -0,0 +1,283 @@
+// License for the specific language governing permissions and limitations under
+// the License.
+object JsonHelpers {
+  implicit val json: EntityDecoder[Json] = org.http4s.jawn.jawnDecoder(facade)
+  def jsonOf[A](implicit decoder: Decoder[A]): EntityDecoder[A] =
+    json.flatMapR { json =>
+      decoder.decodeJson(json).fold(
+        failure =>
+          DecodeResult.failure(InvalidMessageBodyFailure(s"Could not decode JSON: $json", Some(failure))),
+        DecodeResult.success(_)
+      )
+    }
+
+  implicit val jsonEncoder: EntityEncoder[Json] =
+    EntityEncoder[String].contramap[Json] { json =>
+      Printer.noSpaces.pretty(json)
+    }.withContentType(`Content-Type`(MediaType.`application/json`))
+  def jsonEncoderOf[A](implicit encoder: Encoder[A]): EntityEncoder[A] =
+    jsonEncoder.contramap[A](encoder.apply)
+}
+
+object Generators {
+  implicit def arbSession: Arbitrary[Session] =
+    Arbitrary(for {
+      string <- Gen.alphaStr
+      number <- arbitrary[Int]
+    } yield Json.obj("string" -> string.asJson, "number" -> number.asJson))
+}
+
+object Matchers {
+  import org.specs2.matcher.Matchers._
+
+  /** Doesn't capture cookies that are being deleted. */
+  def setCookies(response: Response): Vector[Cookie] =
+    response.headers.collect {
+      case `Set-Cookie`(setCookie) if setCookie.cookie.expires.forall(i => !i.toInstant.isBefore(Instant.now())) =>
+        //        _ >= java.time.Instant.now()) =>
+        setCookie.cookie
+    }.toVector
+
+  def setCookies(mb: MaybeResponse): Vector[Cookie] =
+    mb match {
+      case response: Response =>
+        setCookies(response)
+      case Pass =>
+        Vector.empty
+    }
+
+  def expiredCookies(response: Response): Vector[Cookie] =
+    response.headers.collect {
+      case `Set-Cookie`(setCookie) if setCookie.cookie.expires.exists(_.toInstant.isBefore(Instant.now())) =>
+        setCookie.cookie
+    }.toVector
+
+  def expiredCookies(mb: MaybeResponse): Vector[Cookie] =
+    mb match {
+      case response: Response =>
+        expiredCookies(response)
+      case Pass =>
+        Vector.empty
+    }
+
+  def beCookieWithName(name: String): Matcher[Cookie] =
+    be_===(name) ^^ ((_: Cookie).name)
+
+  def beCookieWhoseContentContains(subcontent: String): Matcher[Cookie] =
+    contain(subcontent) ^^ ((_: Cookie).content)
+
+  def haveSetCookie(cookieName: String): Matcher[Response] =
+    contain(beCookieWithName(cookieName)) ^^ setCookies _
+
+  def haveClearedCookie(cookieName: String): Matcher[Response] =
+    contain(beCookieWithName(cookieName)) ^^ expiredCookies _
+
+  def haveStatus(status: Status): Matcher[Response] =
+    be_===(status) ^^ ((_: Response).status)
+
+  def haveBody[A: EntityDecoder](a: A): Matcher[Response] =
+    TaskMatchers.returnValue(be_===(a)) ^^ ((_: Response).as[A])
+}
+
+//object SessionSpec extends Specification with ScalaCheck with TaskMatchers with ResultMatchers {
```

Comment:
I'm a huge fan of large part of the source code commented out

Reply:
Ah, yes didn't mean to leave those commented out. Fixed now.